src/os/bluestore/BlueFS.cc

```diff
@@ -2475,110 +2638,123 @@ void BlueFS::_rewrite_log_and_layout_sync_LNF_LD(bool allocate_with_fallback,
   }
 #endif
   _flush_bdev();
+
+  // 2.4 Finalize log update
   ++log.seq_live;
   dirty.seq_live = log.seq_live;
   log.t.seq = log.seq_live;
+  vselector->sub_usage(log_file->vselector_hint, old_log_fnode);
+  vselector->add_usage(log_file->vselector_hint, log_file->fnode);
 
-  super.memorized_layout = layout;
-  super.log_fnode = log_file->fnode;
-  // rename device if needed
-  if (log_dev != log_dev_new) {
-    dout(10) << __func__ << " renaming log extents to " << log_dev_new << dendl;
-    for (auto& p : super.log_fnode.extents) {
-      p.bdev = log_dev_new;
-    }
-  }
-  dout(10) << __func__ << " writing super, log fnode: " << super.log_fnode << dendl;
+  // Part 3.
+  // Write out new superblock to reflect all the changes.
+  //
 
+  super.memorized_layout = layout;
   _write_super(super_dev);
   _flush_bdev();
 
-  dout(10) << __func__ << " release old log extents " << old_fnode.extents << dendl;
-  std::lock_guard dl(dirty.lock);
-  for (auto& r : old_fnode.extents) {
-    dirty.pending_release[r.bdev].insert(r.offset, r.length);
+  // we're mostly done
+  dout(10) << __func__ << " log extents " << log_file->fnode.extents << dendl;
+  logger->inc(l_bluefs_log_compactions);
+
+  // Part 4
+  // Finalization. Release old space.
+  //
+  {
+    dout(10) << __func__
+             << " release old log extents " << old_log_fnode.extents
+             << dendl;
+    std::lock_guard dl(dirty.lock);
+    for (auto& r : old_log_fnode.extents) {
+      dirty.pending_release[r.bdev].insert(r.offset, r.length);
+    }
   }
 }
 
 /*
- * 1. Allocate a new extent to continue the log, and then log an event
- * that jumps the log write position to the new extent.  At this point, the
- * old extent(s) won't be written to, and reflect everything to compact.
- * New events will be written to the new region that we'll keep.
+ * ASYNC LOG COMPACTION
  *
- * 2. While still holding the lock, encode a bufferlist that dumps all of the
- * in-memory fnodes and names.  This will become the new beginning of the
- * log.  The last event will jump to the log continuation extent from #1.
- *
- * 3. Queue a write to a new extent for the new beginnging of the log.
+ * 0. Lock the log and forbid its extension. The former covers just
```

Comment:
But I see a possibly to replace `op_update_inc` with `op_update`.
That would allow us to backport to pacific too.
For that we need to fix
`_replay(...) ...  case OP_JUMP:`
since it will fail on backward offset jump.

```
  SUPERBLOCK:
  - contains ino.1 extents of STARTER

  STARTER (seq 1):
  - op_init
  - op_update(ino.1, extents of META_DUMP) <- replaces ino.1 in-place
  - op_jump(2, 0 (beginning of META_DUMP))
  - unused space

  META_DUMP (seq 2):
  - ... dump_metadata
  - op_update(ino.1, extents of LOG_CONTINUATION) <- replaces ino.1
  - op_jump(LOG_CONT.seq, 0, (beginning of LOG_CONT))
  - unused space

  LOG_CONT (seq cont):
  - the continuation of previous log

```

Reply:
@aclamk - honestly I don't fully understand neither your concerns nor the solution. Where does this backward offset jump come from? What's wrong with these two jumps:
op_jump(2, sizeof(STARTER))
...
op_jump(LOG_CONT.seq, sizeof(STARTER) + sizeof(META_DUMP)

could you please elaborate?

And yes I'm planning to backport this to Pacific along with op_update_inc patch. The latter is on its way, see https://github.com/ceph/ceph/pull/48915

Reply:
Sure, if we want to backport op_update_inc too than it is better!